SteamKit2/SteamKit2/Steam/Handlers/SteamMatchmaking/SteamMatchmaking.cs

```diff
@@ -0,0 +1,664 @@
+        /// <param name="lobbySteamId">The SteamID of the lobby we're inviting a user to.</param>
+        /// <param name="userSteamId">The SteamID of the user we're inviting.</param>
+        public void InviteToLobby( uint appId, SteamID lobbySteamId, SteamID userSteamId )
+        {
+            var getLobbyData = new ClientMsgProtobuf<CMsgClientMMSInviteToLobby>( EMsg.ClientMMSInviteToLobby )
+            {
+                Body =
+                {
+                    app_id = appId,
+                    steam_id_lobby = lobbySteamId,
+                    steam_id_user_invited = userSteamId
+                }
+            };
+
+            Send( getLobbyData, appId );
+        }
+
+        /// <summary>
+        /// Obtains a <see cref="Lobby"/>, by its SteamID, if the data is cached locally.
+        /// This method does not send a network request.
+        /// </summary>
+        /// <param name="appId">The ID of app which we're attempting to obtain a lobby for.</param>
+        /// <param name="lobbySteamId">The SteamID of the lobby that should be returned.</param>
+        /// <returns>The <see cref="Lobby"/> corresponding with the specified app and lobby ID, if cached. Otherwise, <c>null</c>.</returns>
+        public Lobby GetLobby( uint appId, SteamID lobbySteamId )
+        {
+            return lobbyCache.GetLobby( appId, lobbySteamId );
+        }
+
+        /// <summary>
+        /// Sends a matchmaking message for a specific app.
+        /// </summary>
+        /// <param name="msg">The matchmaking message to send.</param>
+        /// <param name="appId">The ID of the app this message pertains to.</param>
+        public void Send( ClientMsgProtobuf msg, uint appId )
+        {
+            if ( msg == null )
+            {
+                throw new ArgumentNullException( nameof(msg) );
+            }
+
+            msg.ProtoHeader.routing_appid = appId;
+            Client.Send( msg );
+        }
+
+        /// <summary>
+        /// Handles a client message. This should not be called directly.
+        /// </summary>
+        /// <param name="packetMsg">The packet message that contains the data.</param>
+        public override void HandleMsg( IPacketMsg packetMsg )
+        {
+            if ( packetMsg == null )
+            {
+                throw new ArgumentNullException( nameof(packetMsg) );
+            }
+
+            if ( dispatchMap.TryGetValue( packetMsg.MsgType, out var handler ) )
+            {
+                handler( packetMsg );
+            }
+        }
+
+        internal void ClearLobbyCache()
+        {
+            lobbyCache.Clear();
+        }
+
+        AsyncJob<T> MonitorJob<T>( AsyncJob<T> job )
+            where T : CallbackMsg
+        {
+            job.ToTask().ContinueWith( task =>
+            {
+                lobbyManipulationRequests.TryRemove( job.JobID, out _ );
+            }, TaskContinuationOptions.NotOnRanToCompletion );
```

Comment:
It took me a while to figure out what this is doing and why we're using this particular `TaskContinuationOptions`. Can you leave a comment in here explaining it?

Reply:
If the code's not easy to comprehend it's probably just subpar code. So I'll refactor a smidgen and see if that hits the spot. If that still doesn't do it, let me know and then as a last resort I'll throw in a comment as well.

Reply:
It's probably not the _best_ named method ever, however do you think what I've got now is sufficient?

https://github.com/Benjamin-Dobell/SteamKit/commit/ae8487d47462e1cbffe6ca4d655f59446a439c18

Reply:
I don't think so.

The bit I had trouble with was "okay we'll remove it on faulted or cancelled, but why not on completed?"

Renaming the method to indicate that we handle incomplete outcomes doesn't explain why we don't do it for completed ones.

Reply:
That's fair. Comment added.